admission-webhook/webhook.go

```diff
@@ -0,0 +1,402 @@
+	}
+
+	pod, err := unmarshallPod(request.Object)
+	if err != nil {
+		return nil, err
+	}
+
+	switch request.Operation {
+	case admissionv1beta1.Create:
+		switch operation {
+		case validate:
+			return webhook.validateCreateRequest(pod, request.Namespace)
+		case mutate:
+			return webhook.mutateCreateRequest(pod)
+		default:
+			// shouldn't happen, but needed so that all paths in the function have a return value
+			panic(fmt.Errorf("unexpected webhook operation: %v", operation))
+		}
+
+	case admissionv1beta1.Update:
+		if operation == validate {
+			oldPod, err := unmarshallPod(request.OldObject)
+			if err != nil {
+				return nil, err
+			}
+			return validateUpdateRequest(pod, oldPod)
+		}
+
+		// we only do validation on updates, no mutation
+		return &admissionv1beta1.AdmissionResponse{Allowed: true}, nil
+	default:
+		return nil, &podAdmissionError{error: fmt.Errorf("unpexpected operation %s", request.Operation), pod: pod, code: http.StatusBadRequest}
+	}
+}
+
+// unmarshallPod unmarshalls a pod object from its raw JSON representation.
+func unmarshallPod(object runtime.RawExtension) (*corev1.Pod, *podAdmissionError) {
+	pod := &corev1.Pod{}
+	if err := json.Unmarshal(object.Raw, pod); err != nil {
+		return nil, &podAdmissionError{error: fmt.Errorf("unable to unmarshall pod JSON object: %v", err), code: http.StatusBadRequest}
+	}
+
+	return pod, nil
+}
+
+// validateCreateRequest ensures that the only GMSA content annotations set on the pod,
+// match the corresponding GMSA name annotations, and that the pod's service account
+// is authorized to `use` the requested GMSA's.
+func (webhook *webhook) validateCreateRequest(pod *corev1.Pod, namespace string) (*admissionv1beta1.AdmissionResponse, *podAdmissionError) {
+	var err *podAdmissionError
+
+	iterateOverGMSAAnnotationPairs(pod, func(nameKey, contentsKey string) {
+		if err != nil {
+			return
+		}
+
+		if credSpecName, present := pod.Annotations[nameKey]; present && credSpecName != "" {
+			// let's check that the associated service account can read the relevant cred spec CRD
+			if authorized, reason := webhook.client.isAuthorizedToUseCredSpec(pod.Spec.ServiceAccountName, namespace, credSpecName); !authorized {
+				msg := fmt.Sprintf("service account %s is not authorized `use` gMSA cred spec %s", pod.Spec.ServiceAccountName, credSpecName)
+				if reason != "" {
+					msg += fmt.Sprintf(", reason: %q", reason)
+				}
+				err = &podAdmissionError{error: fmt.Errorf(msg), pod: pod, code: http.StatusForbidden}
+				return
+			}
+
+			// and the content annotation should contain the expected cred spec
+			if credSpecContents, present := pod.Annotations[contentsKey]; present {
+				if expectedContents, code, retrieveErr := webhook.client.retrieveCredSpecContents(credSpecName); retrieveErr != nil {
+					err = &podAdmissionError{error: retrieveErr, pod: pod, code: code}
+				} else if credSpecContents != expectedContents {
```

Comment:
We need a deep equals check of the JSON elements here rather than a string equals check. Note that user may also specify the JSON and that may not be equal from a string comparison perspective but could be equal based on a deep equals/semantic level check. In such cases, the validation should pass.

Reply:
Makes sense indeed, given that as discussed above we want to let users directly set the pod annotations. Amending.